src/qibo/models/dbi/utils_scheduling.py

```diff
@@ -0,0 +1,337 @@
+
+    roots = np.roots(coef)
+    real_positive_roots = [
+        np.real(root) for root in roots if np.imag(root) < error and np.real(root) > 0
+    ]
+    # solution exists, return minimum s
+    if len(real_positive_roots) > 0:
+        sol = min(real_positive_roots)
+        for s in real_positive_roots:
+            if dbi_object.loss(s, d) < dbi_object.loss(sol, d):
+                sol = s
+        return sol
+    # solution does not exist, return None
+    else:
+        return None
+
+
+def off_diagonal_norm_polynomial_expansion_coef(dbi_object, d, n):
+    if d is None:
+        d = dbi_object.diagonal_h_matrix
+    # generate Gamma's where $\Gamma_{k+1}=[W, \Gamma_{k}], $\Gamma_0=H
+    W = dbi_object.commutator(d, dbi_object.sigma(dbi_object.h.matrix))
+    Gamma_list = dbi_object.generate_Gamma_list(n + 2, d)
+    sigma_Gamma_list = list(map(dbi_object.sigma, Gamma_list))
+    exp_list = np.array([1 / math.factorial(k) for k in range(n + 1)])
+    # coefficients for rotation with [W,H] and H
+    c1 = exp_list.reshape((-1, 1, 1)) * sigma_Gamma_list[1:]
+    c2 = exp_list.reshape((-1, 1, 1)) * sigma_Gamma_list[:-1]
+    # product coefficient
+    trace_coefficients = [0] * (2 * n + 1)
+    for k in range(n + 1):
+        for j in range(n + 1):
+            power = k + j
+            product_matrix = c1[k] @ c2[j]
+            trace_coefficients[power] += 2 * np.trace(product_matrix)
+    # coefficients from high to low (n:0)
+    coef = list(reversed(trace_coefficients[: n + 1]))
+    return coef
+
+
+def least_squares_polynomial_expansion_coef(dbi_object, d, n):
+    if d is None:
+        d = dbi_object.diagonal_h_matrix
+    # generate Gamma's where $\Gamma_{k+1}=[W, \Gamma_{k}], $\Gamma_0=H
+    Gamma_list = dbi_object.generate_Gamma_list(n + 1, d)
+    exp_list = np.array([1 / math.factorial(k) for k in range(n + 1)])
+    # coefficients
+    coef = np.empty(n)
+    for i in range(n):
+        coef[i] = np.real(exp_list[i] * np.trace(d @ Gamma_list[i + 1]))
+    coef = list(reversed(coef))
+    return coef
+
+
+# TODO: add a general expansion formula not stopping at 3rd order
+def energy_fluctuation_polynomial_expansion_coef(dbi_object, d, n, state):
+    if d is None:
+        d = dbi_object.diagonal_h_matrix
+    # generate Gamma's where $\Gamma_{k+1}=[W, \Gamma_{k}], $\Gamma_0=H
+    Gamma_list = dbi_object.generate_Gamma_list(n + 1, d)
+    # coefficients
+    coef = np.empty(3)
+    coef[0] = np.real(2 * covariance(Gamma_list[0], Gamma_list[1], state))
+    coef[1] = np.real(2 * variance(Gamma_list[1], state))
+    coef[2] = np.real(
+        covariance(Gamma_list[0], Gamma_list[3], state)
+        + 3 * covariance(Gamma_list[1], Gamma_list[2], state)
+    )
+    coef = list(reversed(coef))
+    return coef
+
+
+def dGamma_diDiagonal(dbi_object, d, H, n, i, dGamma, Gamma_list):
+    # Derivative of gamma with respect to diagonal elements of D (full-diagonal ansatz)
+    A = np.zeros(d.shape)
+    A[i, i] = 1
```

Comment:
```suggestion
    a = np.eye(d.shape[0])
```

@andrea-pasquale can you confirm if all variables should be lower-case letters because of some PEP standards?

Reply:
https://github.com/qiboteam/qibo/pull/1269#discussion_r1537411824